Add EIP: Precompile Falcon512 generic verifier (#8190)
ethereum/EIPs: EIPS/eip-7619.md

## Specification

A clean implementation for Falcon-512 algorithm can be found at [PQClean-Falcon-512](https://github.com/PQClean/PQClean/commit/8e220a87308154d48fdfac40abbb191ac7fce06a)

> The precompiled signature verification function runs at address `0x65`. The required inputs are:

Reviewer: you can put this in a param (precompile address) and refer the param here

Author: I have addressed the observation by adding the implementation for falcon512 into the assets/eip-7619 folder.

---
eip: 7619
title: Precompile Falcon512 generic verifier

> description: Proposal to add precompiled contract that allows generic signature verifications using Falcon512 postquantum algorithm

Reviewer 2: ```suggestion
description: Add precompiled contract that allows generic signature verifications using Falcon512 postquantum algorithm
```

author: Erick Pacheco Pedraza (@eum602), Marcos Allende <mallende@lacnet.com>, Diego Lopez León <dieguitoll@gmail.com>
discussions-to: https://ethereum-magicians.org/t/falcon-512-precompiled-generic-signature-verifier/18569
status: Draft
type: Standards Track
category: Core
created: 2023-07-03
---

## Abstract

> This standard proposes the inclusion of a precompiled signature verification function. The algorithm used is Falcon-512, candidate to be standarized by NIST and quantum resistant with security level I.

Reviewer 2: ```suggestion
Include a precompiled signature verification function using Falcon-512. Falcon-512 is a candidate for standardization by the National Institute of Standards and Technology (NIST) and is quantum resistant with security level I.
```

## Motivation

> The advent of quantum computing threatens blockchain protocols and networks because they utilize non-quantum resistant cryptographic algorithms. When quantum computers become robust enough to run Shor’s algorithm on a large scale, the most used asymmetric algorithms, utilized for digital signatures and message encryption, such as RSA, (EC)DSA, and (EC)DH, will be no longer secure. Quantum computers will be able to break them within a short period of time. Today, there are hundreds of billions of dollars denominated in cryptocurrencies and other digital assets that rely on blockchain ledgers as well as thousands of blockchain-based applications storing value in blockchain networks. Cryptocurrencies and blockchain-based applications require solutions that guarantee quantum resistance in order to preserve the integrity of data and assets in these public and immutable ledgers. Most proposals for quantum-resistant blockchain networks are theoretical, require large QKD networks, or propose new quantum-resistant blockchain protocols to be built from scratch. This EIP is pioneer in proposing a solution compatible with current EVM blockchain protocols. It presents a simple mechanism to add a NIST-compliant post-quantum signature to blockchain transactions, making them quantum-resistant even when ECC cryptography becomes vulnerable against attacks by quantum computers. We have developed a Solidity implementation for the on-chain verification of this signatures, which does not scale due to the required high amount of gas. This is why **this EIP is proposing a pre-compiled smart contract** that allows to verify post-quantum signatures in a scalable manner.

Reviewer 3: Shor (This algorithm is like a super fast method for quantum computers to find the prime factors of large numbers. While it has great potential, it is not yet ready to be used with large numbers because quantum computers need more reliability, but they are continually getting better)
RSA, (EC)DSA and (EC)DH (Those algorithms are based on easily reversible trapdoor functions, for a quantum computer when those computers become more reliable and accessible)

## Specification

The precompiled signature verification function runs at address `0x65`. The required inputs are:

- `message` - an arbitrary nummber of bytes representing the message that was signed
- `public key` - Falcon Public key of 897 bytes
- `signature` - 666 bytes (max size)

Those inputs are encoded according to:

```
[4 bytes method signature][32 bytes signature offset][32 bytes public key offset][32 bytes message offset][32 bytes signature length][bytes signature][32 bytes public key length][bytes public key][32 bytes message length][bytes message]
```

> 

Reviewer 2: Please include a source (see [EIP-1](https://eips.ethereum.org/EIPS/eip-1#linking-to-external-resources) for permitted sources) for the Falcon-512 algorithm. Otherwise I don't think this EIP has enough detail to be implementable.

Author: Hi @gaudren thanks for the feedback, I have addressed all observations. Since I have added an external source for Falcon-512 I am getting an error. Waiting for next steps and fixes, if needed.

Author: Adding a new PR to include post quantum external source links in EIP-1 in this PR(https://github.com/ethereum/EIPs/pull/8245) @gaudren @abcoathup

### Example Usage in Solidity

The precompile can be wrapped in Solidity to provide a more development-friendly interface to `Falcon-512` signature verification.

```solidity
  function verify(
        bytes calldata signature,
        bytes calldata publicKey,
        bytes calldata message,
        address falconVerifier
    ) public returns (bool isValid) {
        (bool success, bytes memory verifies) = address(0x0000000000000000000000000000000000000065)
            .staticcall(
                abi.encodeWithSignature(
                    "verify(bytes,bytes,bytes)",
                    signature,
                    publicKey,
                    message
                )
            );
        require(success && verifies.length == 32, "Invalid signature");
        return verifies[31] == 0;
    }
```

### Gas costs and benchmarks

Since data size is variable, each operation to execute the precompiled call will require:

- 1465 base units of gas for each operation.
- 6 units of gas for each for each word contained in the payload.
  These values are based on data presented at the benchmark section.

## Rationale

> Falcon-512 is a good candidate for a new signature implementation. First of all, key sizes are relatively small compared to ther signature algorithms that are also quantum resistant.

Reviewer 2: > ther

Maybe their or the?

> The following table shows a summary of the typicall values for keys and the approximate amount of time taken per second to execute signature verifications as specified by the authors of Falcon algorithm:

Reviewer 2: ```suggestion
The following table shows a summary of the typical values for keys and the approximate amount of time taken per second to execute signature verifications as specified by the authors of Falcon algorithm:
```

Implementing just the signature verification is sufficiently enough to open a wide range of use cases at the smart contract level. To give an example, this method can perfectly work with the Account Abstraction concept, where the authentication process would rely on Falcon and the account is a smart contract.

## Backwards Compatibility

There is very little risk of breaking backwards-compatibility with this EIP, the sole issue being if someone were to build a contract relying on the address at `0x65` being empty. The likelihood of this is low, and should specific instances arise, the address could be chosen to be any arbitrary value with negligible risk of collision.

## Test Cases

- A set of test vectors, with valid signatures, for benchmarking on new implementations is located in a separate [file](../assets/eip-7619/bench_vectors.md)

- A set of test vectors, with invalid data, to verify some invalid signature cases is located in a separate [file](../assets/eip-7619/invalid_signature_test_vectors.md)

> - Additionally when method signature doesn't match with `de8f50a1` or just payload length is zero the precompile will throw an exception:

Reviewer 2: This line is introducing a requirement so it should be in the Specification section.

```
  {
		Input:         "",
		ExpectedError: "invalid input format",
		Name:          "vector 0: empty input",
	},
	{
		Input:         "111111110000000000000000000000000000000000000000000000000000000000000060000000000000000000000000000000000000000000000000000000000000032000000000000000000000000000000000000000000000000000000000000006e0",
		ExpectedError: "invalid method signature",
		Name:          "vector 1: 4 method signature bytes is not de8f50a1",
	}
```

## Reference Implementation

> - The reference implementation for the precompiled signature using Falcon can be found for the Go-Ethereum client and for Hyperledger Besu.

Reviewer 2: External links aren't permitted. This section is optional, so you may remove it entirely if you'd like.

> ### benchmarks

Reviewer 2: Since these benchmarks are used to calculate the gas cost, they should be in the Rationale section.

Assuming ecRecover precompile is perfectly priced, a set of benchmarks were executed comparing Falcon-512 signature verification function precompile with ecRecover precompile. For benchmarks, it was used 2.6 GHz Intel Core i7 64-bit machine.

```sh
$ cat /proc/cpuinfo | grep "model name"
Intel(R) Core(TM) i7-10750H CPU @ 2.60GHz
```

#### 5 seconds benchmark

Based on [100 test vectors with valid signatures](../assets/eip-7619/bench_vectors.md) a benchmark of 5 seconds, for each test vector, was performed, considering the following:

- the average time per operation (ns/op),
- each verctor size increases in 33 bytes
- Ecrecover is well tested

Then the 100 BenchmarkPrecompiledFalcon512 results were approximated to a linear equation dependent on the number of bytes in each test vector. From the aproximation it was concluded that 1465 units of gas is the base requirement for any falcon-512 signature validation and 6 gas per word is additionally required. These results were summarized in the `Gas costs and benchmarks` section.

> ```

Reviewer 2: This is a large table, and should be included as an asset.
